### PR TITLE
[CI] Test building for 32-bit arch

### LINKED ISSUE #10015: 32-bit arch support and WASM
Hi, I'm one of the core developers of [Pyodide](https://github.com/pyodide/pyodide), which is porting the Python ecosystem to the browser using WebAssembly.
First of all, thank you for maintaining this project for a long time!

In Pyodide, we have been providing WebAssembly version of xgboost since 2022. WebAssembly is based on the 32bit architecture, so we've been having some trouble updating xgboost since it decided to drop 32bit arch support.

As an open-source maintainer, I understand the challenges of managing a project with a small resource, so I fully understand the decision to remove support for 32bit systems to reduce management costs. However, given that the use of Python in WebAssembly environments may become more prevalent in the future (at least I hope), it would be very happy if we could discuss how xgboost can support 32-bit arch or WASM in the long term.

### PR DESCRIPTION
**Pull request by hcho3**
Closes #8964
Closes #10015 

* Attempt a minimal change to code base to eliminate narrowing errors from the compiler.
* Also experiment with `docker/build-push-action`, which massively simplifies caching of Docker containers. It transparently uses GitHub's built-in caching feature to cache containers.
* The 32-bit Debian container ships with latest GCC (13) which throws warning for overloaded virtual functions. For now, this warning would not be considered an error.

### CODE REVIEW

File: src/common/hist_util.h [hunk 1]
```diff
@@ -113,7 +113,7 @@ class HistogramCuts {
     auto end = ptrs[column_id + 1];
     auto beg = ptrs[column_id];
     auto it = std::upper_bound(values.cbegin() + beg, values.cbegin() + end, value);
-    auto idx = it - values.cbegin();
+    auto idx = static_cast<uint32_t>(it - values.cbegin());
```
Comment (trivialfis) [inline src/common/hist_util.h:116]:
```suggestion
    auto idx = static_cast<bst_bin_t>(it - values.cbegin());
```

File: src/common/hist_util.h [hunk 2]
```diff
@@ -136,7 +136,7 @@ class HistogramCuts {
     auto beg = ptrs[fidx] + vals.cbegin();
     // Truncates the value in case it's not perfectly rounded.
     auto v = static_cast<float>(common::AsCat(value));
-    auto bin_idx = std::lower_bound(beg, end, v) - vals.cbegin();
+    auto bin_idx = static_cast<uint32_t>(std::lower_bound(beg, end, v) - vals.cbegin());
```
Comment (trivialfis) [inline src/common/hist_util.h:139]:
```suggestion
    auto bin_idx = static_cast<bst_bin_t>(std::lower_bound(beg, end, v) - vals.cbegin());
```

**Review by trivialfis — APPROVED** [review]
Thank you for working on the 32-bit build!

We will probably need to have some follow-up PRs for running some basic tests in C++, but there is no need to rush. Feel free to merge as you see fit.

**Comment by barracuda156** [comment]
@hcho3 Thank you very much for working on this!

@trivialfis Will this be included into the next CRAN release? (I recall I could not simply apply patches from the master onto R package code.)

**Comment by hcho3** [comment]
> Will this be included into the next CRAN release?

Probably not, since we are already having trouble uploading XGBoost 2.0 to CRAN. We are currently discussing next steps, see #9810.
In the meanwhile, you are free to build XGBoost R package from the source. (Use `R CMD INSTALL .` from the `R-package` directory.)

**Comment by barracuda156** [comment]
@hcho3 Thanks, then I can set it up manually with cloning the repo to run tests. (I need Macports environment for `R` stuff to work correctly.) Will update on results.

**Pull request merged**